File: Day219/longestSubstingWithKUniqueCharacters.go
package main

import (
	"fmt"
)

func longetsSubstring(str string, k int ) int {
	if len(str) == 0 || k == 0 {
		return 0
	}
	if (k>len(str)) {
		return -1
	}
	startIndex, endIndex := 0, 0
	answer := 0
	memory := make(map[byte]int)
	for endIndex < len(str) {
		if  _,is := memory[str[endIndex]]; !is{
			k--
		}
		memory[str[endIndex]] = endIndex
		for k < 0 {
			if memory[str[startIndex]] == startIndex {
				k++
				delete(memory, str[startIndex])
			}
			startIndex++
		}
		answer = max(answer, endIndex - startIndex + 1)
		endIndex++
	}
	return answer
}

func main() {
	str := "aabacbghjyelllloooobebe";
	k := 3;
	fmt.Println(longetsSubstring(str , k))
}

File: Day219/longestSubstingWithKUniqueCharacters.java
package Day219;

import java.util.Map;
import java.util.HashMap;

public class longestSubstingWithKUniqueCharacters {
    
    public static int longestSubstring(String str, int k) {
        if (k>str.length()) {
            return -1;
        }
        if (str.length() == 0 || k == 0) {
            return 0;
        }
        int startIndex = 0, endIndex = 0;
        Map<Character, Integer> map = new HashMap<>();
        int answer = 0;
        while (endIndex < str.length()) {
            if (!map.containsKey(str.charAt(endIndex))) {
                k--;
            }
            map.put(str.charAt(endIndex), endIndex);
            while (k < 0) {
                if (map.get(str.charAt(startIndex)) == startIndex) {
                    k++;
                    map.remove(str.charAt(startIndex));
                }
                startIndex++;
            }
            answer = Math.max(answer, endIndex - startIndex + 1);
            endIndex++;
        }
        return answer;
    }

    public static void main(String[] args) {
        String s = "aabacbghjyelllloooobebe";
        int k = 3;
        System.out.println(longestSubstring(s, k));
    }
}